uint8 id
uint8 code
uint8 comp_type
uint8 comp_id